// @1inch/fusion-sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderStatus = exports.Quote = exports.Preset = exports.PresetEnum = exports.QuoterCustomPresetRequest = exports.RelayerRequest = exports.QuoterRequest = exports.MakerTraits = exports.LimitOrderContract = exports.getLimitOrderV4Domain = exports.AmountMode = exports.ExtensionBuilder = exports.TakerTraits = exports.Interaction = exports.getLimitOrderContract = exports.randBigInt = exports.Extension = exports.Address = void 0;
var limit_order_sdk_1 = require("@1inch/limit-order-sdk");
Object.defineProperty(exports, "Address", { enumerable: true, get: function () { return limit_order_sdk_1.Address; } });
Object.defineProperty(exports, "Extension", { enumerable: true, get: function () { return limit_order_sdk_1.Extension; } });
Object.defineProperty(exports, "randBigInt", { enumerable: true, get: function () { return limit_order_sdk_1.randBigInt; } });
Object.defineProperty(exports, "getLimitOrderContract", { enumerable: true, get: function () { return limit_order_sdk_1.getLimitOrderContract; } });
Object.defineProperty(exports, "Interaction", { enumerable: true, get: function () { return limit_order_sdk_1.Interaction; } });
Object.defineProperty(exports, "TakerTraits", { enumerable: true, get: function () { return limit_order_sdk_1.TakerTraits; } });
Object.defineProperty(exports, "ExtensionBuilder", { enumerable: true, get: function () { return limit_order_sdk_1.ExtensionBuilder; } });
Object.defineProperty(exports, "AmountMode", { enumerable: true, get: function () { return limit_order_sdk_1.AmountMode; } });
Object.defineProperty(exports, "getLimitOrderV4Domain", { enumerable: true, get: function () { return limit_order_sdk_1.getLimitOrderV4Domain; } });
Object.defineProperty(exports, "LimitOrderContract", { enumerable: true, get: function () { return limit_order_sdk_1.LimitOrderContract; } });
Object.defineProperty(exports, "MakerTraits", { enumerable: true, get: function () { return limit_order_sdk_1.MakerTraits; } });
__exportStar(require("./fusion-order/index"), exports);
__exportStar(require("./auction-calculator/index"), exports);
__exportStar(require("./connector/index"), exports);
__exportStar(require("./sdk/index"), exports);
__exportStar(require("./constants"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./utils/amounts"), exports);
__exportStar(require("./utils/time"), exports);
__exportStar(require("./validations"), exports);
__exportStar(require("./ws-api"), exports);
__exportStar(require("./errors"), exports);
var api_1 = require("./api");
Object.defineProperty(exports, "QuoterRequest", { enumerable: true, get: function () { return api_1.QuoterRequest; } });
Object.defineProperty(exports, "RelayerRequest", { enumerable: true, get: function () { return api_1.RelayerRequest; } });
Object.defineProperty(exports, "QuoterCustomPresetRequest", { enumerable: true, get: function () { return api_1.QuoterCustomPresetRequest; } });
Object.defineProperty(exports, "PresetEnum", { enumerable: true, get: function () { return api_1.PresetEnum; } });
Object.defineProperty(exports, "Preset", { enumerable: true, get: function () { return api_1.Preset; } });
Object.defineProperty(exports, "Quote", { enumerable: true, get: function () { return api_1.Quote; } });
Object.defineProperty(exports, "OrderStatus", { enumerable: true, get: function () { return api_1.OrderStatus; } });
