// @1inch/fusion-sdk/dist/fusion-order/fusion-extension.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const limit_order_sdk_1 = require("@1inch/limit-order-sdk");
const fusion_order_1 = require("./fusion-order");
const auction_details_1 = require("./auction-details");
const fusion_extension_1 = require("./fusion-extension");
describe('FusionExtension', () => {
    it('should decode', () => {
        const extensionContract = new limit_order_sdk_1.Address('0x8273f37417da37c4a6c3995e82cf442f87a25d9c');
        const order = fusion_order_1.FusionOrder.new(extensionContract, {
            makerAsset: new limit_order_sdk_1.Address('0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'),
            takerAsset: new limit_order_sdk_1.Address('0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'),
            makingAmount: 1000000000000000000n,
            takingAmount: 1420000000n,
            maker: new limit_order_sdk_1.Address('0x00000000219ab540356cbb839cbe05303d7705fa'),
            salt: 10n
        }, {
            auction: new auction_details_1.AuctionDetails({
                duration: 180n,
                startTime: 1673548149n,
                initialRateBump: 50000,
                points: [
                    {
                        coefficient: 20000,
                        delay: 12
                    }
                ]
            }),
            whitelist: [
                {
                    address: new limit_order_sdk_1.Address('0x00000000219ab540356cbb839cbe05303d7705fa'),
                    allowFrom: 0n
                }
            ],
            resolvingStartTime: 1673548139n
        });
        const fusionExtension = fusion_extension_1.FusionExtension.decode(order.extension.encode());
        expect(fusionExtension).toStrictEqual(order.fusionExtension);
    });
    it('should decode with permit', () => {
        const extensionContract = new limit_order_sdk_1.Address('0x8273f37417da37c4a6c3995e82cf442f87a25d9c');
        const order = fusion_order_1.FusionOrder.new(extensionContract, {
            makerAsset: new limit_order_sdk_1.Address('0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'),
            takerAsset: new limit_order_sdk_1.Address('0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48'),
            makingAmount: 1000000000000000000n,
            takingAmount: 1420000000n,
            maker: new limit_order_sdk_1.Address('0x00000000219ab540356cbb839cbe05303d7705fa'),
            salt: 10n
        }, {
            auction: new auction_details_1.AuctionDetails({
                duration: 180n,
                startTime: 1673548149n,
                initialRateBump: 50000,
                points: [
                    {
                        coefficient: 20000,
                        delay: 12
                    }
                ]
            }),
            whitelist: [
                {
                    address: new limit_order_sdk_1.Address('0x00000000219ab540356cbb839cbe05303d7705fa'),
                    allowFrom: 0n
                }
            ],
            resolvingStartTime: 1673548139n
        }, {
            permit: '0xdeadbeef'
        });
        const fusionExtension = fusion_extension_1.FusionExtension.decode(order.extension.encode());
        expect(fusionExtension).toStrictEqual(order.fusionExtension);
    });
});
